### PR TITLE
Oauth middleware

### PR DESCRIPTION
**Pull request by JorritSalverda**

### CODE REVIEW

File: auth/middleware.go [hunk 1]
```diff
@@ -4,8 +4,11 @@ import (
 	"fmt"
 	"net/http"
 	"strings"
+	"time"
 
+	jwt "github.com/appleboy/gin-jwt/v2"
```
Comment (vitorarins) [inline auth/middleware.go:9]: Just a quick heads up, I was checking https://jwt.io/ and it seems the underlying library being used by `gin-jwt` doesn't do all checks that JWT provides.
The website has some alternatives but then I don't know about using Gin with them.

File: config/config.go [hunk 2]
```diff
@@ -58,8 +67,144 @@ func (c *APIServerConfig) ReadLogFromCloudStorage() bool {
...
+			"user.read+openid+profile+email",
+		}
+	case "facebook":
+		oauthConfig.Endpoint = endpoints.Facebook
+	case "github":
+		oauthConfig.Endpoint = endpoints.Facebook
+	case "bitbucket":
+		oauthConfig.Endpoint = endpoints.Facebook
+	case "gitlab":
+		oauthConfig.Endpoint = endpoints.Facebook
```
Comment (vitorarins) [inline config/config.go:120]: not sure if this was intended, but hey are all set to Facebook

File: services/rbac/service.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
+package rbac
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
```
Comment (vitorarins) [inline services/rbac/service.go:9]: I don't know how up to date is the https://jwt.io/. Maybe this library already implemented those checks. But would nice to check.
Also, we are checking token agains database/provider's, right?

Comment (JorritSalverda) [reply services/rbac/service.go:9]: There's 2 things, we go through an oauth grant for gaining access to a users details (from Google, Microsoft, etc) and then trust that this person is who he/she says she is (effectively externalizing the identity trust) and from there on generate our own JWT to secure communications with the Estafette API.

What the jwt-go library is lacking is at least an option to set clock skew. And we can set a couple of extra claims and check those in an incoming token, for example the `issued at time` and the `issuer`: https://github.com/estafette/estafette-ci-api/blob/master/auth/helpers.go#L180-L182

Currently there's a check to see if the token is expired, but that's all: https://github.com/appleboy/gin-jwt/blob/master/auth_jwt.go#L373-L386

File: config/config.go (hunk 2 above)
Comment (JorritSalverda) [reply config/config.go:120]: Error, but they haven't been implemented correctly yet anyway; I need to figure out for all of them where to get the user profile and implement that and map the response to our `UserIdentity` object.

File: auth/middleware.go (hunk 1 above)
Comment (JorritSalverda) [reply auth/middleware.go:9]: Here's the current checks: https://github.com/appleboy/gin-jwt/blob/master/auth_jwt.go#L373-L386. We can use the `IdentityHandler` and/or `Authorizator` for additional checks.

**Pull request merged**